benchmarks/opmphm.c

```diff
@@ -980,6 +989,787 @@ static void benchmarkMappingAllSeeds (void)
+}
+
+/**
+ * END ==================================================== OPMPHM Search Time ========================================================= END
+ */
+
+/**
+ * START ================================================= Binary search Time ======================================================== START
+ *
+ * This benchmark measures the time of the binary search.
+ * Uses all KeySet shapes, for one n (KeySet size) ksPerN KeySets are used.
+ * Each measurement done with one KeySet is repeated numberOfRepeats time and summarized with the median.
+ * For one n (KeySet size) the ksPerN results are also summarized with the median.
+ * The results are written out in the following format:
+ *
+ * n;search_1;search_2;...;search_(numberOfSearches)
+ *
+ * The number of needed seeds for this benchmarks is: numberOfShapes * nCount * ksPerN * (1  + searchesCount )
+ */
+
+static void benchmarkBinarySearchTime (void)
+{
+	benchmarkSearchTime ("benchmark_binary_search_time", KDB_O_NOCASCADING);
+}
+
+/**
+ * END =================================================== Binary search Time ========================================================== END
+ */
+
+/**
+ * START ================================================= hsearch Build Time ======================================================== START
+ *
+ * This benchmark measures the time of the hsearch build.
+ * For one n (KeySet size) ksPerN KeySets are used, with different loads.
+ * This benchmark has a 10 strike policy, when 10 time the measured time is over 10000 the next KeySet shape is handled.
+ * The results are written out in the following format:
+ *
+ * n;ks;load;time
+ *
+ * The number of needed seeds for this benchmarks is: numberOfShapesUsed * nCount * ksPerN
+ */
+
+// where does hsearch stores the strings??? no out of mem but speed loosing
+
+// clang-format off
+// format bug
+#ifdef HAVE_HSEARCHR
+// clang-format on
+
+/**
+ * @brief Measures the hsearch build numberOfRepeats time and returns median
+ *
+ * @param ks the KeySet
+ * @param nI the KeySet size
+ * @param load the load
+ * @param repeats array to store repeated measurements
+ * @param numberOfRepeats fields in repeats
+ *
+ * @retval median time
+ */
+static size_t benchmarkHsearchBuildTimeMeasure (KeySet * ks, size_t nI, double load, size_t * repeats, size_t numberOfRepeats)
+{
+	for (size_t repeatsI = 0; repeatsI < numberOfRepeats; ++repeatsI)
+	{
+		// preparation for measurement
+		struct timeval start;
+		struct timeval end;
+		Key * key;
+		ksRewind (ks);
+		ENTRY e;
+		ENTRY * ep;
+		// fresh htab
+		struct hsearch_data * htab = elektraCalloc (sizeof (struct hsearch_data));
+		if (!htab)
+		{
+			printExit ("calloc");
+		}
+
+		// START MEASUREMENT
+		__asm__("");
+		gettimeofday (&start, 0);
+		__asm__("");
+
+		if (!hcreate_r (nI / load, htab))
+		{
+			printExit ("hcreate_r");
+		}
+
+		while ((key = ksNext (ks)))
+		{
+			e.key = (char *) keyName (key);
+			e.data = key;
+			if (!hsearch_r (e, ENTER, &ep, htab))
```

Comment:
Here we search for every key in a keyset? What about searching for keys not in a keyset?

Reply:
This is only a sanity check, a compare between hsearch search binary search search and hsearch search OPMPHM search was not planned.